src/cpr_data_access/models.py

```diff
@@ -771,6 +822,118 @@ def add_spans(
+        :return self:
+        """
+
+        if target_model not in {CPRDocument, GSTDocument}:
+            raise ValueError("target_model must be one of {CPRDocument, GSTDocument}")
+
+        # Raises ValueError if metadata CSV doesn't contain the required columns
+        metadata_df = _load_and_validate_metadata_csv(metadata_csv_path, target_model)
+
+        new_documents = []
+
+        for document in self.documents:
+            if document.document_id not in metadata_df["CPR Document ID"].tolist():
+                if force_all_documents_have_metadata:
+                    raise Exception(
+                        f"No document exists in the scraper data with ID equal to the document's: {document.document_id}"
+                    )
+                else:
+                    continue
+
+            doc_dict = document.dict(
+                exclude={"document_metadata", "_text_block_idx_hash_map"}
+            )
+            new_metadata_dict = metadata_df.loc[
+                metadata_df["CPR Document ID"] == document.document_id
+            ].to_dict(orient="records")[0]
+
+            if target_model == CPRDocument:
+                doc_metadata = CPRDocumentMetadata(
+                    source="CPR",
+                    geography=new_metadata_dict.pop("Geography"),
+                    geography_iso=new_metadata_dict.pop("Geography ISO"),
+                    slug=new_metadata_dict["CPR Document Slug"],
+                    category=new_metadata_dict.pop("Category"),
+                    type=new_metadata_dict.pop("Document Type"),
+                    sectors=[
+                        s.strip()
+                        for s in new_metadata_dict.get("Sectors", "").split(";")
+                    ],
+                    status=new_metadata_dict.pop("CPR Document Status"),
+                    collection_id=new_metadata_dict.pop("CPR Collection ID"),
+                    collection_name=new_metadata_dict.pop("Collection name"),
+                    family_id=new_metadata_dict.pop("CPR Family ID"),
+                    family_name=new_metadata_dict.pop("Family name"),
+                    family_slug=new_metadata_dict.pop("CPR Family Slug"),
+                    role=new_metadata_dict.pop("Document role"),
+                    variant=new_metadata_dict.pop("Document variant"),
+                    # NOTE: we incorrectly use the "publication_ts" value from the parser output rather than the correct
+                    # document date (calculated from events in product). When we upgrade to Vespa we should use the correct
+                    # date.
+                    publication_ts=document.document_metadata.publication_ts,
+                )
+
+                metadata_at_cpr_document_root = {
```

Comment:
So are we using the `pop` method to remove fields from the dict as we only want a set number in the `document_root`

Reply:
`pop` here removes the fields but we don't actually return the remaining fields in the dict anywhere. i've forgotten why i used `pop` rather than just `['field']` to be honest